src/paulscode/android/mupen64plusae/input/map/PlayerMap.java

```diff
@@ -125,15 +187,89 @@ public void removeUnavailableMappings()
             if( !AbstractProvider.isHardwareAvailable( id ) )
             {
                 Log.v( "PlayerMap", "Removing device " + id + " from map" );
-                mMap.removeAt( i );
+                unmap( id );
             }
         }
     }
     
+    /**
+     * Serializes the map data to a string.
+     * 
+     * @return The string representation of the map data.
+     */
+    @Override
+    public String serialize()
+    {
+        // Serialize the map data to a multi-delimited string
+        String result = "";
+        
+        for( Entry<String, Integer> entry : deviceNameToId.entrySet() )
+        {
+            // If the device name is the id, then store the latest id, otherwise the device's unique name
+            String device = ( Integer.getInteger( entry.getKey(), 0 ) != 0 ) ?
+                    entry.getValue().toString() :
+                    entry.getKey();
+            int player = mMap.get( entry.getValue() );
+            
+            // Putting the value first makes the string a bit more human readable IMO
+            if( player > 0 && device != null )
+                result += player + ":" + device + ",";
+        }
+        
+        return result;
+    }
+    
+    /**
+     * Deserializes the map data from a string.
+     * 
+     * @param s The string representation of the map data.
+     */
     @Override
     public void deserialize( String s )
     {
-        super.deserialize( s );
-        removeUnavailableMappings();
+        // Reset the map
+        unmapAll();
+        
+        // Parse the new map data from the multi-delimited string
+        if( s != null )
+        {
+            int psuedoId = -100;
+            
+            // Read the input mappings
+            String[] pairs = s.split( "," );
+            for( String pair : pairs )
+            {
+                String[] elements = pair.split( ":" );
+                if( elements.length == 2 )
+                {
+                    try
+                    {
+                        int player = Integer.parseInt( elements[0] );
+                        int id = AbstractProvider.getHardwareId( elements[1] );
+                        
+                        // If the specified device is not currently connected.
+                        if( id == 0 )
+                        {
+                            // Do not add disconnected devices if the key specified is a device id.
+                            if( Integer.getInteger( elements[1], 0 ) != 0 )
```

Comment:
This (and all other references to Integer.getInteger) should be Integer.parseInt. Integer.getInteger does not do what I though it did. I think this may be the source of all the old device ids showing up even though they are disconnected. I will fix this.